stdlib/public/SwiftShims/DispatchOverlayShims.h

```diff
@@ -0,0 +1,213 @@
+}
+
+static inline dispatch_queue_t
+_swift_dispatch_get_main_queue(void) {
+  return dispatch_get_main_queue();
+}
+
+static inline dispatch_data_t
+_swift_dispatch_data_empty(void) {
+  return dispatch_data_empty;
+}
+
+static inline __swift_shims_dispatch_block_t _Nullable
+_swift_dispatch_data_destructor_default(void) {
+  return DISPATCH_DATA_DESTRUCTOR_DEFAULT;
+}
+
+static inline __swift_shims_dispatch_block_t
+_swift_dispatch_data_destructor_free(void) {
+  return _dispatch_data_destructor_free;
+}
+
+static inline __swift_shims_dispatch_block_t
+_swift_dispatch_data_destructor_munmap(void) {
+  return _dispatch_data_destructor_munmap;
+}
+
+#define SWIFT_DISPATCH_SOURCE(t)                                               \
+  static inline dispatch_source_type_t _swift_dispatch_source_type_##t(void) { \
+    return DISPATCH_SOURCE_TYPE_##t;                                           \
+  }
+
+SWIFT_DISPATCH_SOURCE(DATA_ADD)
+SWIFT_DISPATCH_SOURCE(DATA_OR)
+SWIFT_DISPATCH_SOURCE(MACH_SEND)
+SWIFT_DISPATCH_SOURCE(MACH_RECV)
+SWIFT_DISPATCH_SOURCE(MEMORYPRESSURE)
+SWIFT_DISPATCH_SOURCE(PROC)
+SWIFT_DISPATCH_SOURCE(READ)
+SWIFT_DISPATCH_SOURCE(SIGNAL)
+SWIFT_DISPATCH_SOURCE(TIMER)
+SWIFT_DISPATCH_SOURCE(VNODE)
+SWIFT_DISPATCH_SOURCE(WRITE)
+
+
+static inline __swift_shims_dispatch_block_t
+_swift_dispatch_block_create_with_qos_class(
+    dispatch_block_flags_t flags,
+    qos_class_t qos,
+    int relative_priority,
+    __swift_shims_dispatch_block_t _Nonnull block) {
+  return dispatch_block_create_with_qos_class(
+      flags, qos, relative_priority, block);
+}
+
+static inline __swift_shims_dispatch_block_t
+_swift_dispatch_block_create_noescape(
+    dispatch_block_flags_t flags,
+    __swift_shims_dispatch_block_t SWIFT_DISPATCH_NOESCAPE block) {
+  return dispatch_block_create(flags, block);
+}
+
+static inline int _swift_dispatch_block_wait(
+    __swift_shims_dispatch_block_t block,
+    unsigned long long timeout) {
+  return dispatch_block_wait(block, timeout);
+}
+
+static inline void _swift_dispatch_block_notify(
+    __swift_shims_dispatch_block_t block,
+    dispatch_queue_t queue,
+    __swift_shims_dispatch_block_t notifier) {
+  dispatch_block_notify(block, queue, notifier);
+}
+
+static inline void _swift_dispatch_block_cancel(
+    __swift_shims_dispatch_block_t block) {
+  dispatch_block_cancel(block);
+}
+
+static inline int _swift_dispatch_block_testcancel(
+    __swift_shims_dispatch_block_t block) {
+  return dispatch_block_testcancel(block);
+}
+
+static inline void _swift_dispatch_async(
+    dispatch_queue_t queue,
+    __swift_shims_dispatch_block_t block) {
+  dispatch_async(queue, block);
+}
+
+static inline void _swift_dispatch_sync(
+    dispatch_queue_t queue,
+    __swift_shims_dispatch_block_t block) {
+  dispatch_sync(queue, block);
+}
+
+static inline void _swift_dispatch_barrier_async(
+    dispatch_queue_t queue,
+    __swift_shims_dispatch_block_t block) {
+  dispatch_barrier_async(queue, block);
+}
+
```

Comment:
We don't have a `sync<T>` call that takes a `DispatchWorkItem`. The reason we have to have these shims is to preserve the identity of `DispatchWorkItem/dispatch_block_t` such that the runtime doesn't wrap them in thunk closures. The barrier case in `sync<T>` is only ever fed closures anyway, so we don't need a shim here.